src/Microsoft.ML.FastTree/TreeEnsemble/QuantileRegressionTree.cs

```diff
@@ -91,6 +97,39 @@ public void SetLabelsDistribution(double[] labelsDistribution, double[] weights)
             _instanceWeights = weights;
         }
 
+        /// <summary>
+        /// Copy training examples' labels and their weights to external variables.
+        /// </summary>
+        /// <param name="leafSamples">List of label collections. The type of a collction is a double array. The i-th label collection contains training examples' labels falling into the i-th leaf.</param>
+        /// <param name="leafSampleWeights">List of labels' weight collections. The type of a collction is a double array. The i-th collection contains weights of labels falling into the i-th leaf.
+        /// Specifically, leafSampleWeights[i][j] is the weight of leafSamples[i][j].</param>
+        internal void ExtractLeafSamplesAndTheirWeights(out double[][] leafSamples, out double[][] leafSampleWeights)
+        {
+            leafSamples = new double[NumLeaves][];
+            leafSampleWeights = new double[NumLeaves][];
+            // If there is no training labels stored, we view the i-th leaf value as the only label stored at the i-th leaf.
+            var sampleCountPerLeaf = _labelsDistribution != null ? _labelsDistribution.Length / NumLeaves : 1;
+            for (int i = 0; i < NumLeaves; ++i)
+            {
+                var samplesPerLeaf = new List<double>();
+                var sampleWeightsPerLeaf = new List<double>();
+                for (int j = 0; j < sampleCountPerLeaf; ++j)
```

Comment:
I think you can make them as an array and pre allocate before you enter the loop.

Reply:
No problem.

---
In reply to: [252027300](https://github.com/dotnet/machinelearning/pull/2243#discussion_r252027300) [](ancestors = 252027300)